### PR TITLE
Markdown

### PR DESCRIPTION
**Pull request by benjamin-kirkbride**
Accidentally closed https://github.com/Akuli/porcupine/pull/1314 :sweat_smile:

### CODE REVIEW

**Comment by Akuli** [comment]
As for git messes: it's totally ok to make a mess with git! I use "squash and merge" in my projects, which basically means that a messy PR will appear as just one tidy commit when someone looks at the Git history.

You might also like https://akuli.github.io/git-guide/

File: porcupine/plugins/markdown.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import re
+import tkinter
+from functools import partial
+
+from porcupine import get_tab_manager, tabs, textutils, utils
+
+log = logging.getLogger(__name__)
+
+
+setup_before = ["tabs2spaces", "autoindent"]
```
Comment (Akuli) [inline porcupine/plugins/markdown.py:18]: Maybe must be also be before autocomplete, so that `return "break"` prevents autocompletions?

You can try running Porcupine with `--shuffle-plugins` a few times to see if you have defined all necessary dependencies. (See `--help`)

File: tests/test_indent_dedent.py [hunk 2]
```diff
@@ -198,28 +198,6 @@ def check(filename, input_commands, output):
     return check
 
 
-def test_markdown_autoindent(check_autoindents):
-    check_autoindents(
```
Comment (Akuli) [inline tests/test_indent_dedent.py:202]: Can we keep this, but just modify it so that it passes?

File: tests/test_markdown_plugin.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    ListItemCase(id="duplicate token 5", line="  -#) item -- 1 -", expected=False),
+    ListItemCase(id="duplicate token 6", line="  *-#)1. item -- 1 -", expected=False),
+]
+
+# test `#` and 0 to 99 numbered lists
+# tests ol with `.` and `)`
+IS_LIST_ITEM_CASES.extend(
+    [
+        ListItemCase(id=f"numbered {i}", line=f"{i}{sep} item 1", expected=True)
+        for i, sep in itertools.product(itertools.chain(range(100), "#"), (".", ")"))
```
Comment (Akuli) [inline tests/test_markdown_plugin.py:46]: These are maybe a bit too thorough :) We really don't need more than a thousand green dots when running pytest because of one regex. But more importantly, these two lines aren't very readable. It took me a while to figure out what they do.

Maybe instead add a few specific cases, so that it's easier to understand?

```
ListItemCase(id=f"numbered 1", line=f"1. item 1", expected=True)
ListItemCase(id=f"numbered 1", line=f"1) item 1", expected=True)
ListItemCase(id=f"numbered 4", line=f"4. item 1", expected=True)
ListItemCase(id=f"numbered 4", line=f"4) item 1", expected=True)
ListItemCase(id=f"numbered 42", line=f"42. item 1", expected=True)
ListItemCase(id=f"numbered 42", line=f"42) item 1", expected=True)
ListItemCase(id=f"numbered #", line=f"#. item 1", expected=True)
ListItemCase(id=f"numbered #", line=f"#) item 1", expected=True)
```

Also applies to other `.extend()`s.

**Review by Akuli — COMMENTED** [review]
I like it overall :)

**Comment by benjamin-kirkbride** [comment]
@Akuli does the error in CI mean anything to you?

File: porcupine/plugins/markdown.py (hunk 1 above)
Comment (benjamin-kirkbride) [reply porcupine/plugins/markdown.py:18]: The tests run with `--shuffle-plugins`, so shouldn't that be covered?

This actually doesn't interfere with autocomplete, the regex doesn't match useless there is a space after the list prefix, so regardless we should be good I believe. If you think that's wrong, or can think of a test to verify that, I'm all ears :)

File: tests/test_indent_dedent.py (hunk 2 above)
Comment (benjamin-kirkbride) [reply tests/test_indent_dedent.py:202]: I got rid of it because it felt redundant with all the markdown tests, but I can look back and see if there is a way to make that happen

File: tests/test_markdown_plugin.py (hunk 3 above)
Comment (benjamin-kirkbride) [reply tests/test_markdown_plugin.py:46]: > We really don't need more than a thousand green dots when running pytest because of one regex

But it's so fun :weary: 

> these two lines aren't very readable

That's fair. It felt readable at the time :sweat: 

> Maybe instead add a few specific cases, so that it's easier to understand?

Yeah I can do this.

On a more broad note, how would you feel about adding https://hypothesis.readthedocs.io/en/latest/ to the test suite?

**Comment by Akuli** [comment]
The CI error has to do with #1300. Apparently the CI on Python 3.11 doesn't pass very reliably, even though you got it to pass most of the time in #1327.

I'd suggest we do nothing about the CI error if it disappears when tests run again. We will "soon" switch to pyright anyway.

File: tests/test_markdown_plugin.py (hunk 3 above)
Comment (Akuli) [reply tests/test_markdown_plugin.py:46]: I'm not a big fan of `hypothesis`. IMO tests should be as concrete/specific as possible, such as `ListItemCase(id="numbered 42", line="42) item 1", expected=True)`, whereas the code being tested tends to be written very generally. This ensures that at least the test is correct, because it's ideally very simple.

And yes, I know that the point of `hypothesis` is to catch bugs where a specific string that you didn't think of causes a problem. But they tend to get caught without unit tests in Porcupine. For example, the empty `line.splitlines()` bug caused errors when I opened a file.

I can see `hypothesis` being useful in other projects though. For example, maybe you have functions to escape and unescape a string in some way, and you want to know that they work in all corner cases.

Comment (benjamin-kirkbride) [reply tests/test_markdown_plugin.py:46]: I would see hypothesis as purely additive, it wouldn't replace the simple `ListItemCase(id="numbered 42", line="42) item 1", expected=True)`, sanity type tests, but it would replace:

```python
IS_LIST_ITEM_CASES.extend(
    [
        ListItemCase(id=f"numbered {i}", line=f"{i}{sep} item 1", expected=True)
        for i, sep in itertools.product(itertools.chain(range(100), "#"), (".", ")"))
```

> For example, the empty line.splitlines() bug caused errors when I opened a file.

I think there is likely many more of these sorts of things, especially as more and more plugins get added ;)

I think hypothesis would likely help a lot with the flaky tests as well. I should probably move this discussion to an issue.

**Comment by benjamin-kirkbride** [comment]
This PR is blocked by https://github.com/Akuli/porcupine/pull/1349

**Comment by benjamin-kirkbride** [comment]
I am now going to work to finish this using Actions
